paddle/framework/threadpool.h

```diff
@@ -0,0 +1,139 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+#include <condition_variable>
+#include <cstdio>
+#include <functional>
+#include <iostream>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace paddle {
+namespace framework {
+
+class ThreadPool {
+ private:
+  typedef std::function<void()> Task;
+  int num_threads_;
+  int available_;
+  bool running_;
+  bool done_;
+  std::queue<Task> tasks_;
+  std::vector<std::unique_ptr<std::thread>> threads_;
+  std::mutex mutex_;
+  std::condition_variable scheduled_;
+  std::condition_variable completed_;
+
+ public:
+  /**
+   * @brief   get a instance of threadpool, the thread number will
+   *          be specified by the first time.
+   * @param[in] num_threads   The available thread number.
+   *            If num_threads <= 0, the thread pool wil be initilized
+   *            with the number of concurrent threads supported.
+   */
+  static ThreadPool* Instance(size_t num_threads) {
```

Comment:
Maybe just bake in `num_threads` (not as argument, `std::thread::hardware_concurrency` is probably a good choice. But it could return 0, so need to check for 0), otherwise the first time `Instance(10)` is called, latter `Instance(20)` is called, but returns one threadpool with 10 threads rather than 20.

Reply:
The user may specify the max CPU cores which will be used, such as [GOMAXPROCS](https://docs.google.com/document/d/1At2Ls5_fhJQ59kDK2DFVhFu3g5mATSXqqV5QrxinasI/edit) in Go, but maybe I can remove the argument `num_threads` and support `MAXPROCS` for future.

Reply:
Done.